sormas-ui/src/main/java/de/symeda/sormas/ui/utils/BulkOperationHandler.java

```diff
@@ -0,0 +1,168 @@
+package de.symeda.sormas.ui.utils;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import com.vaadin.server.Sizeable;
+import com.vaadin.ui.Label;
+import com.vaadin.ui.Notification;
+import com.vaadin.ui.UI;
+import com.vaadin.ui.Window;
+
+import de.symeda.sormas.api.i18n.Captions;
+import de.symeda.sormas.api.i18n.I18nProperties;
+import de.symeda.sormas.api.i18n.Strings;
+import de.symeda.sormas.api.uuid.HasUuid;
+import de.symeda.sormas.ui.utils.components.progress.BulkProgressLayout;
+import de.symeda.sormas.ui.utils.components.progress.BulkProgressUpdateInfo;
+import de.symeda.sormas.ui.utils.components.progress.ProgressResult;
+
+public class BulkOperationHandler {
+
+	/**
+	 * Amount of DTOs that are forwarded to the backend in one single call
+	 * when displaying a progress layout.
+	 */
+	public static final int BULK_ACTION_BATCH_SIZE = 20;
+	/**
+	 * Amount of DTOs that have to be selected for the progress layout to be displayed.
+	 */
+	public static final int BULK_ACTION_PROGRESS_THRESHOLD = 40;
+	private boolean cancelAfterCurrentBatch;
+	private boolean cancelButtonClicked;
+	private final Lock cancelLock = new ReentrantLock();
+	private int initialEntryCount;
+	private int successfulEntryCount;
+
+	public void doBulkOperation(
+		Function<List<? extends HasUuid>, Integer> bulkOperationFunction,
+		List<? extends HasUuid> selectedEntries,
+		Consumer<List<? extends HasUuid>> bulkOperationDoneCallback) {
+
+		initialEntryCount = selectedEntries.size();
+		if (selectedEntries.size() < BULK_ACTION_PROGRESS_THRESHOLD) {
+			successfulEntryCount = bulkOperationFunction.apply(selectedEntries);
+			if (initialEntryCount > successfulEntryCount) {
+				NotificationHelper.showNotification(
+					String.format(I18nProperties.getString(Strings.messageEntriesEditedExceptArchived), successfulEntryCount),
+					Notification.Type.HUMANIZED_MESSAGE,
+					-1);
+			} else {
+				NotificationHelper.showNotification(I18nProperties.getString(Strings.messageEntriesEdited), Notification.Type.HUMANIZED_MESSAGE, -1);
+			}
+			bulkOperationDoneCallback.accept(Collections.emptyList());
+		} else {
+			UI currentUI = UI.getCurrent();
+
+			BulkProgressLayout bulkProgressLayout = new BulkProgressLayout(currentUI, selectedEntries.size(), this::handleCancelButtonClicked);
+			Window window = VaadinUiUtil.createPopupWindow();
+			window.setCaption(I18nProperties.getString(Strings.headingDataImport));
```

Comment:
the data import header should be replaced with something more suitable for Bulk Edit

Reply:
Aah, great, I've already created that string, but forgot to use it :D